File: _recipes/node/retry.ts
// Recipe: exponential-backoff retry helper
// Stdlib-only.
// Usage:
//   import { retry, RetryError } from './retry.js';
//
//   const data = await retry(() => fetch(url), { maxAttempts: 5, baseDelayMs: 100 });
//
// Errors that aren't instances of any class in `only` propagate immediately.
// After maxAttempts failures the last caught error is re-thrown wrapped in RetryError.

export class RetryError extends Error {
  constructor(message: string, readonly cause: unknown) {
    super(message);
    this.name = 'RetryError';
  }
}

export interface RetryOptions {
  maxAttempts?: number;
  baseDelayMs?: number;
  maxDelayMs?: number;
  jitter?: number;
  /** Error classes to catch and retry. Anything else propagates. */
  only?: Array<abstract new (...args: never[]) => Error>;
}

const sleep = (ms: number): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve, ms));

export async function retry<T>(
  fn: () => Promise<T>,
  options: RetryOptions = {},
): Promise<T> {
  const {
    maxAttempts = 3,
    baseDelayMs = 100,
    maxDelayMs = 30_000,
    jitter = 0.1,
    only = [Error],
  } = options;

  if (maxAttempts < 1) {
    throw new Error('maxAttempts must be >= 1');
  }

  let lastErr: unknown;
  for (let attempt = 0; attempt < maxAttempts; attempt += 1) {
    try {
      return await fn();
    } catch (err) {
      const matches = only.some((Klass) => err instanceof Klass);
      if (!matches) throw err;
      lastErr = err;
      if (attempt === maxAttempts - 1) break;
      let delay = Math.min(baseDelayMs * 2 ** attempt, maxDelayMs);
      if (jitter > 0) delay += Math.random() * delay * jitter;
      await sleep(delay);
    }
  }
  throw new RetryError(`retried ${maxAttempts} times`, lastErr);
}

File: _recipes/python/retry.py
# Recipe: exponential-backoff retry decorator
# Stdlib-only.
# Usage:
#   from retry import retry, RetryError
#
#   @retry(max_attempts=5, base_delay=0.1)
#   def fetch_or_die() -> bytes:
#       ...
#
# Exceptions outside `only=` propagate immediately. After max_attempts
# failures the last caught exception is re-raised wrapped in RetryError.
from __future__ import annotations

import random
import time
from functools import wraps
from typing import Any, Callable, Tuple, Type, TypeVar

F = TypeVar("F", bound=Callable[..., Any])


class RetryError(RuntimeError):
    """Raised when all retry attempts have been exhausted."""


def retry(
    *,
    max_attempts: int = 3,
    base_delay: float = 0.1,
    max_delay: float = 30.0,
    jitter: float = 0.1,
    only: Tuple[Type[BaseException], ...] = (Exception,),
) -> Callable[[F], F]:
    """Return a decorator that retries the wrapped function with exponential backoff.

    Args:
        max_attempts: total tries including the first. Must be >= 1.
        base_delay: seconds to wait before the second attempt. Each subsequent
            attempt doubles the delay, capped at max_delay.
        max_delay: cap on per-attempt sleep in seconds.
        jitter: max random jitter added per sleep, as a fraction of the delay
            (0.0 disables jitter).
        only: exception types to catch. Anything not matching propagates.
    """
    if max_attempts < 1:
        raise ValueError("max_attempts must be >= 1")

    def decorator(fn: F) -> F:
        @wraps(fn)
        def wrapper(*args: Any, **kwargs: Any) -> Any:
            last_exc: BaseException | None = None
            for attempt in range(max_attempts):
                try:
                    return fn(*args, **kwargs)
                except only as exc:  # noqa: B030
                    last_exc = exc
                    if attempt == max_attempts - 1:
                        break
                    delay = min(base_delay * (2 ** attempt), max_delay)
                    if jitter > 0:
                        delay += random.random() * delay * jitter
                    time.sleep(delay)
            raise RetryError(f"{fn.__name__} failed after {max_attempts} attempts") from last_exc

        return wrapper  # type: ignore[return-value]

    return decorator
